dFusion/dFusionSpec.md

```diff
@@ -79,10 +79,93 @@ To summarize, here is a list of state that is stored inside the smart contract:
+    - there is no account currently occupying the requested index and
+    - the sender does not currently occupy an account slot.
+
+On the level of contract storage, these contraints imply a bijective mapping {1, 2, ..., K} <-> addresses.
+
+Note that: Registering accounts by specified index (rather than incrementing) enables the possiblity for accounts to be closed and account slots to be made available.
+
+#### Registering tokens
+
+Token registration is performed by token address.
+Tokens are indexed incrementally (implying that they can never be removed) from 1 to `T` (the maximum number of allowed tokens).
+Only the contract owner may register tokens by address.
+
+**Note that** the contract does not verify the requested token address is actually that of an ERC20 token. This implies that security is enforced by permissions (i.e. only owner may register tokens).
+
+In the event that a *non-token* is registered as a token, the indexing slot would be permanently occupied, but no successful deposits could ever be made. However, if the registered token does implement a transferFrom successful deposits could occur
+
+### Deposits & Withdrawals
+
+#### Depositing token
+
+Deposits are requested by token index and amount.
+The contract verifies, that
+        - the sender has registered an account,
+        - the token index is that of a valid registered token and
+        - the sender has sufficient balance
+
+Token transfers (from sender to contract) are made during the deposit request phase, but the sender's balance is not represented in the `stateRoot` until these requests have been processed.
+
+Sender may not exit their balances untill the stateRoot has been updated with the appropriate deposit requests. However, deposit processing will be made publicially available. That is to say, any end user could potentially process their own deposits.
+
+Upon successful transfer, the deposit is included in the appropriate depositRequest slot and the EVM emit's a "Deposit" event containing the following information
+    - Account ID,
+    - Token Index,
+    - Amount Transferred,
+    - Deposit Slot
+
+Where, deposit slot is deterministically governed the EVM's current block number as the integer division of block number by 20. This allows for asynchronicity so that one knows (after a certain block) that the deposit hash is no longer `active` (i.e. will not change). This is required for the asynchronous handling of in-flight transactions.
+
+#### Processing Deposits:
+
+Deposits may be applied by specifying deposit slot and updated `stateRoot`. This new state root is computed by
+- gathering all the deposit events for that slot,
+- computing the updated balances for all cooresponding deposit transactions and
+- hashing this information together (as a Pedersen Hash)
+
+For security reasons, the `applyDeposits` function must be called with the following parameters
+- slot,
+- currentStateHash,
+- currentDepositHash and
```

Comment:
Can you explain why you need the depositHash and the slot? Wouldn't we be able to look up the depositHash from the slot? Given that we require slots to be frozen, I can't think of a race condition we would need to defend against.

Also, "current" is not really clear, since there might be a lot of depositHashes existing in parallel. What exactly do you mean by "current"? Is it mutated as a result of this function?

Reply:
This is only an additional security layer, to ensure that the sender knows which deposit hash they are intending to change. 

Yes, I can change current to simply depositHash

Reply:
👍 I think we don't need to send redundant information unless it can change in some form of race condition.

Reply:
So then, if I understand correctly, I will remove the depositHash verification from apply deposits and update this part of the spec to reflect it.